### PR TITLE
Added raw 'deflate' (no wrap) support to JZlibDecoder

### PR DESCRIPTION
**Pull request by MediumOne**
Motivation:

Some servers use raw 'deflate' (no wrap) encoding instead of Zlib.
JdkZlibDecoder has support to detect this and fallback to raw
'deflate' encoding when the wrapper type is ZLIB_OR_NONE.
However JZlibDecoder does not have this support and fails when
raw 'deflate' encoding is used.

Modifications:

Modified JZlibDecoder to add this "fallback" feature similar to
JdkZlibDecoder when the wrapper type is ZLIB_OR_NONE. Before this
change, JZlibDecoder was able to handle gzip compression when
wrapper is ZLIB_OR_NONE. This has been removed now to make the
behavior same as JdkZlibDecoder. Since this is removed now,
I have removed the related test case - testZLIB_OR_NONE3().

Result:

JZlibDecoder now supports fallback to raw deflate (no wrap)
encoding similar to JdkZlibDecoder.

### CODE REVIEW

**Comment by normanmaurer** [comment]
Looking

File: codec/src/test/java/io/netty/handler/codec/compression/JZlibTest.java [hunk 1]
```diff
@@ -26,4 +28,10 @@ protected ZlibEncoder createEncoder(ZlibWrapper wrapper) {
     protected ZlibDecoder createDecoder(ZlibWrapper wrapper) {
         return new JZlibDecoder(wrapper);
     }
+
+    @Test(expected = DecompressionException.class)
+    @Override
+    public void testZLIB_OR_NONE3() throws Exception {
```
Comment (normanmaurer) [inline codec/src/test/java/io/netty/handler/codec/compression/JZlibTest.java]: @MediumOne why this change ?

File: codec/src/test/java/io/netty/handler/codec/compression/ZlibCrossTest1.java [hunk 2]
```diff
@@ -26,4 +28,10 @@ protected ZlibEncoder createEncoder(ZlibWrapper wrapper) {
     protected ZlibDecoder createDecoder(ZlibWrapper wrapper) {
         return new JZlibDecoder(wrapper);
     }
+
+    @Test(expected = DecompressionException.class)
+    @Override
```
Comment (normanmaurer) [inline codec/src/test/java/io/netty/handler/codec/compression/ZlibCrossTest1.java]: @MediumOne why this change ?

File: codec/src/main/java/io/netty/handler/codec/compression/JZlibDecoder.java [hunk 3]
```diff
@@ -95,6 +104,24 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) t
...
+            boolean zlib = ZlibUtil.looksLikeZlib(in.getShort(in.readerIndex()));
+
+            JZlib.WrapperType wrapperType = zlib ? JZlib.W_ZLIB : JZlib.W_NONE;
+
+            int resultCode = z.init(wrapperType);
+            if (resultCode != JZlib.Z_OK) {
+                ZlibUtil.fail(z, "initialization failure", resultCode);
+            }
+
+            decideZlibOrNone = false;
```
Comment (Scottmitch) [inline codec/src/main/java/io/netty/handler/codec/compression/JZlibDecoder.java]: put this in a `finally { .. }` block?

File: codec/src/test/java/io/netty/handler/codec/compression/JZlibTest.java (hunk 1 above)
Comment (Scottmitch) [inline codec/src/test/java/io/netty/handler/codec/compression/JZlibTest.java]: I guess to expect the `DecompressionException` exception?

Comment (MediumOne) [inline codec/src/test/java/io/netty/handler/codec/compression/JZlibTest.java]: This is a mistake. I simply copied the behavior from `JdkZlibTest` without realizing that JZlib library handles gzip decompression when the `ZlibWrapper` type is `ZLIB_OR_NONE` (which is mapped to `JZlib.W_ANY` and this handles gzip).

However when I tried to fix this, I realized that `JdkZlibDecoder` does not handle gzip decompression when the Wrapper type is `ZLIB_OR_NONE`. Only `JZlibDecoder` does this as it finally maps `ZLIB_OR_NONE` to JZlib's `JZlib.W_ANY` which handles gzip. Is it okay if `ZLIB_OR_NONE` handles only zlib or none and not gzip? In this case, we can remove the test case `testZLIB_OR_NONE3()` altogether. 

If `JZlibDecoder` should handle gzip for `ZLIB_OR_NONE`, then then I can detect the compression format in decode (using a new method like `ZlibUtil.looksLikeGzip()`) and then set the JZlib's wrapper type as `W_ANY`.

Please let me know your thoughts and I will modify the PR.

Comment (normanmaurer) [inline codec/src/test/java/io/netty/handler/codec/compression/JZlibTest.java]: Both should work the same way, so a fix would be awesome!

Comment (MediumOne) [inline codec/src/test/java/io/netty/handler/codec/compression/JZlibTest.java]: Both work the same way with the current changes (before this they didn't!). Both do not handle gzip compression when the JZlibdecoder and JdkZlibDecoders are used with ZLIB_OR_NONE wrapper type. Is this behavior okay?

The other way is to make both these decoders handle gzip compression when the wrapper type is ZLIB_OR_NONE. 

Please let me know which way to choose.

Comment (normanmaurer) [inline codec/src/test/java/io/netty/handler/codec/compression/JZlibTest.java]: as long as both work the same way it's fine I think

Comment (MediumOne) [inline codec/src/test/java/io/netty/handler/codec/compression/JZlibTest.java]: Okay. I will leave the feature changes as they are currently. I don't see why gzip should be handled when the wrapper type is ZLIB_OR_NONE. I will also remove the test case - testZLIB_OR_NONE3() as it tests a feature which is not implemented.

**Comment by normanmaurer** [comment]
@MediumOne let me know once you are done and I will have another look

File: codec/src/main/java/io/netty/handler/codec/compression/JZlibDecoder.java (hunk 3 above)
Comment (MediumOne) [inline codec/src/main/java/io/netty/handler/codec/compression/JZlibDecoder.java]: Added. Thanks.

**Comment by MediumOne** [comment]
@normanmaurer @Scottmitch - I worked on the review comments and I have pushed the changes. PTAL.

File: codec/src/test/java/io/netty/handler/codec/compression/JZlibTest.java [hunk 4]
```diff
@@ -15,6 +15,8 @@
  */
 package io.netty.handler.codec.compression;
 
+import org.junit.Test;
```
Comment (Scottmitch) [inline codec/src/test/java/io/netty/handler/codec/compression/JZlibTest.java]: revert?

File: codec/src/test/java/io/netty/handler/codec/compression/ZlibCrossTest1.java [hunk 5]
```diff
@@ -15,6 +15,8 @@
  */
 package io.netty.handler.codec.compression;
 
+import org.junit.Test;
```
Comment (Scottmitch) [inline codec/src/test/java/io/netty/handler/codec/compression/ZlibCrossTest1.java]: revert?

File: codec/src/main/java/io/netty/handler/codec/compression/JZlibDecoder.java [hunk 6]
```diff
@@ -95,6 +104,26 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) t
             return;
         }
 
+        if (decideZlibOrNone) {
+            // First two bytes are needed to decide if it's a ZLIB stream.
+            if (in.readableBytes() < 2) {
```
Comment (Scottmitch) [inline codec/src/main/java/io/netty/handler/codec/compression/JZlibDecoder.java]: use `inputLength` which was saved above?

**Comment by Scottmitch** [comment]
@MediumOne - Few more comments and LGTM. Is is possible to create a unit test for what you have fixed?

**Comment by MediumOne** [comment]
There is already a unit test for raw deflate encoding (testZLIB_OR_NONE). So JZlib has been detecting and handling raw deflate encoding when the wrapper type is set to JZlib.W_ANY. For some reason, my specific use case fails (of using raw deflate encoding) when this wrapper type is set to JZlib.W_ANY. It works only after the changes in this PR. I will decline this PR until I understand the root cause and then make changes if necessary. Thanks for the review comments.

**Comment by Scottmitch** [comment]
@MediumOne - Let us know if you have any questions or need help.
